tests/index/test_index_bifm.cpp

```diff
@@ -0,0 +1,200 @@
+//       notice, this list of conditions and the following disclaimer.
+//     * Redistributions in binary form must reproduce the above copyright
+//       notice, this list of conditions and the following disclaimer in the
+//       documentation and/or other materials provided with the distribution.
+//     * Neither the name of Knut Reinert or the FU Berlin nor the names of
+//       its contributors may be used to endorse or promote products derived
+//       from this software without specific prior written permission.
+//
+// THIS SOFTWARE IS PROVIDED BY THE COPYRIGHT HOLDERS AND CONTRIBUTORS "AS IS"
+// AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE
+// IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE
+// ARE DISCLAIMED. IN NO EVENT SHALL KNUT REINERT OR THE FU BERLIN BE LIABLE
+// FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL
+// DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR
+// SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER
+// CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT
+// LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY
+// OUT OF THE USE OF THIS SOFTWARE, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH
+// DAMAGE.
+//
+// ==========================================================================
+// Author: Enrico Siragusa <enrico.siragusa@fu-berlin.de>
+// ==========================================================================
+
+#include <seqan/basic.h>
+#include <seqan/index.h>
+
+#include "test_index_helpers.h"
+
+using namespace seqan;
+
+template <typename TBiFMIndex>
+inline bool
+testBidirectionalIndex(unsigned int textLength, unsigned int patternLength)
+{
+	typedef DnaString											TText;
+
+	typedef Index<TText, FMIndex<> >							TFMIndex;
+	typedef typename Iterator<TFMIndex, TopDown<> >::Type		TFMIter;
+	typedef typename Iterator<TBiFMIndex, TopDown<> >::Type		TBiFMIter;
+
+	TText text, pattern;
+	generateText<TText>(text, textLength);
+	generateText<TText>(pattern, patternLength);
+
+	TText revText(text);
+	reverse(revText);
+
+	TFMIndex fmIndex1(text);
+	TFMIndex fmIndex2(revText);
+	TFMIter fm1(fmIndex1);
+	TFMIter fm2(fmIndex2);
+
+	TBiFMIndex bifmIndex(text);
+
+	TBiFMIter bifm1(bifmIndex);
+	TBiFMIter bifm2(bifmIndex);
+
+	bool res1 = true, res2 = true, res3 = true, res4 = true;
+
+	unsigned int i;
+	for (i = 0; i < patternLength; ++i)
+	{
+		res1 &= goDown(fm1, pattern[patternLength - i - 1]);
+		res2 &= goDown(fm2, pattern[i]);
+		res3 &= rightExtend(bifm1, pattern[i]);
+		res4 &= leftExtend(bifm2, pattern[patternLength - i - 1]);
+	}
+
+	if (i == patternLength)
+	{
+	    SEQAN_ASSERT_EQ(res1, res2);
+	    SEQAN_ASSERT_EQ(res1, res3);
+	    SEQAN_ASSERT_EQ(res1, res4);
+
+	    if (res1) // if all are true
+	    {
+	    	SEQAN_ASSERT_EQ(fm1.vDesc.range, bifm1.fwdIter.vDesc.range);
```

Comment:
Use the public member function ```range(it)``` instead of accessing the supposedly private member ```it.vDesc```.